### PR TITLE
[WIP] First small optimisation for z-order curve (a<x and x<b => a<x<b)

### PR DESCRIPTION
**Pull request by stavrolia**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.
First small improvement of KeyCondition.
Category (leave one):
- Improvement

### CODE REVIEW

File: dbms/src/Storages/MergeTree/KeyCondition.cpp [hunk 1]
```diff
@@ -444,6 +444,45 @@ void KeyCondition::traverseAST(const ASTPtr & node, const Context & context, Blo
...
                     rpn.emplace_back(std::move(element));
+
+                /* Combine statements like (x >= a) and (x <= b) into x in range [a, b] */
+                if (rpn.size() >= 3 && rpn.back().function == RPNElement::FUNCTION_AND)
+                {
+                    const size_t to_modify = rpn.size() - 3;
+                    const auto rhs = rpn[to_modify + 1];
+                    const auto lhs = rpn[to_modify];
+                    if (
+                            lhs.function == RPNElement::FUNCTION_IN_RANGE &&
```
Comment (akuzm) [inline dbms/src/Storages/MergeTree/KeyCondition.cpp]: Let's remove this line break.

**Comment by akuzm** [comment]
Can we add some kind of test for this code?

File: dbms/src/Storages/MergeTree/KeyCondition.cpp [hunk 2]
```diff
@@ -444,6 +444,45 @@ void KeyCondition::traverseAST(const ASTPtr & node, const Context & context, Blo
                   */
                 if (i != 0 || element.function == RPNElement::FUNCTION_NOT)
                     rpn.emplace_back(std::move(element));
+
+                /* Combine statements like (x >= a) and (x <= b) into x in range [a, b] */
+                if (rpn.size() >= 3 && rpn.back().function == RPNElement::FUNCTION_AND)
+                {
+                    const size_t to_modify = rpn.size() - 3;
+                    const auto rhs = rpn[to_modify + 1];
+                    const auto lhs = rpn[to_modify];
```
Comment (akuzm) [inline dbms/src/Storages/MergeTree/KeyCondition.cpp]: We can use references here.

Also, let's create a merged element on stack and then put it in `rpn` if the merge succeeds -- when modifying in place, we can leave some rubbish, especially if some field is added in the future.

File: dbms/src/Storages/MergeTree/KeyCondition.cpp [hunk 3]
```diff
@@ -444,6 +444,45 @@ void KeyCondition::traverseAST(const ASTPtr & node, const Context & context, Blo
...
+                    if (
+                            lhs.function == RPNElement::FUNCTION_IN_RANGE &&
+                            rhs.function == RPNElement::FUNCTION_IN_RANGE &&
+                            lhs.key_column == rhs.key_column &&
+                            lhs.monotonic_functions_chain == rhs.monotonic_functions_chain)
+                    {
+                        bool merged = false;
+                        if (!lhs.range.right_bounded && !rhs.range.left_bounded)
```
Comment (akuzm) [inline dbms/src/Storages/MergeTree/KeyCondition.cpp]: Can we just always use `Range::intersectRange`, or is there some case when this doesn't work?

**Comment by alexey-milovidov** [comment]
Let's continue?

**Comment by stavrolia** [comment]
Yep, I would soon

**Comment by stavrolia** [comment]
I couldn't reproduce improvement from this optimization and stucked

**Pull request closed without merging**